compilers/concrete-compiler/compiler/lib/Dialect/RT/Transforms/HoistAwaitFuturePass.cpp

```diff
@@ -0,0 +1,259 @@
+// Part of the Concrete Compiler Project, under the BSD3 License with Zama
+// Exceptions. See
+// https://github.com/zama-ai/concrete-compiler-internal/blob/main/LICENSE.txt
+// for license information.
+
+#include <concretelang/Analysis/StaticLoops.h>
+#include <concretelang/Dialect/RT/IR/RTDialect.h>
+#include <concretelang/Dialect/RT/IR/RTOps.h>
+#include <concretelang/Dialect/RT/Transforms/Passes.h>
+
+#include <mlir/Dialect/Utils/StaticValueUtils.h>
+
+#include <iterator>
+#include <optional>
+
+namespace {
+struct HoistAwaitFuturePass
+    : public HoistAwaitFuturePassBase<HoistAwaitFuturePass> {
+  // Checks if all values of `a` are sizes of a non-dynamic dimensions
+  bool allStatic(llvm::ArrayRef<int64_t> a) {
+    return llvm::all_of(
+        a, [](int64_t r) { return !mlir::ShapedType::isDynamic(r); });
+  }
+
+  void runOnOperation() override {
+    mlir::func::FuncOp func = getOperation();
+
+    llvm::SmallVector<mlir::Operation *> opsToErase;
+
+    func.walk([&](mlir::concretelang::RT::AwaitFutureOp awaitFutureOp) {
+      // Make sure there are no other consumers that rely on the
+      // synchronization
+      if (!awaitFutureOp.getResult().hasOneUse())
+        return;
+
+      mlir::scf::ForallOp forallOp =
+          llvm::dyn_cast<mlir::scf::ForallOp>(awaitFutureOp->getParentOp());
+
+      if (!forallOp)
+        return;
+
+      mlir::tensor::ParallelInsertSliceOp parallelInsertSliceOp =
+          llvm::dyn_cast<mlir::tensor::ParallelInsertSliceOp>(
+              awaitFutureOp.getResult().getUses().begin()->getOwner());
+
+      if (!parallelInsertSliceOp)
+        return;
+
+      // Make sure that the original tensor into which the
+      // synchronized values are inserted is a region out argument of
+      // the forall op and thus being written to concurrently
+      mlir::Value dst = parallelInsertSliceOp.getDest();
+
+      if (!llvm::any_of(forallOp.getRegionOutArgs(),
+                        [=](mlir::Value output) { return output == dst; }))
+        return;
+
+      // Currently, the tensor storing the futures must have a static
```

Comment:
Is that tilling could generate non static loops?

Reply:
No, but the pass cannot handle dynamic bounds and there are no guarantees about the IR that serves as an input to the pass.